File: Java/src/com/dbc/algorithm/LinkedList/RotateList.java
package com.dbc.algorithm.LinkedList;

public class RotateList {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return head;
        }

        ListNode front = head;
        ListNode pre = new ListNode(0, head);
        int count = 1;
        while (front.next != null) {
            count++;
            front = front.next;
        }
        front.next = head;
        count = count - (k % count);
        for (int i = 0; i < count; i++) {
            pre = pre.next;
            head = head.next;
        }

        pre.next = null;
        return head;
    }
}

File: src/tmp/Python/linked-list/rotate_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def rotateRight(head: ListNode, k: int) -> ListNode:
    if head is None:
        return head

    pre = ListNode(0, head)
    front = head

    count = 1
    while front.next:
        count += 1
        front = front.next
    
    count = count - (k % count)
    front.next = head
    for i in range(count):
        head = head.next
        pre = pre.next
    
    pre.next = None
    return head

    
